### PR TITLE
Plumb RPC listener up to caller

### PR DESCRIPTION
**Pull request by nickvikeras**
# Description

This PR allows the RPC server's socket address to be returned when initializing the server. This allows the library consumer to easily programmatically determine which port the RPC server is listening on. My use case for this is automated testing. I'd like to be able to simply specify that the server bind to port '0' and then test against whatever port the OS assigns dynamically. I will have many RPC servers running in parallel across many tests within a single process, and I don't want to have to deal with port conflicts. 

## Integration

Integration is straightforward. My main concern is that I am making non-backwards-compatible changes to public library functions. Let me know if I should leave backwards-compatible wrappers in place for any/all of the public functions that were modified. 

## Review Notes
The rationale for making the new listen_addresses field on the RpcHandlers struct a ```[MultiAddr]``` rather than ```SocketAddr``` is because I wanted it to be transport-agnostic as well as capable of supporting multiple listening addresses in case that is ever required by the RPC server in the future. 

# Checklist

* [x] My PR includes a detailed description as outlined in the "Description" and its two subsections above.
* [ ] My PR follows the [labeling requirements](CONTRIBUTING.md#Process) of this project (at minimum one label for `T`
  required)
    * External contributors: ask maintainers to put the right label on your PR.
* [x] I have made corresponding changes to the documentation (if applicable)
* [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)

1. I didn't understand what the 'T' label meant. Am I supposed to open a github Issue for my PR?
2. I didn't see an easy way to add tests since the functions I am modifying are not directly called by any tests.

### CODE REVIEW

File: substrate/client/rpc-servers/src/lib.rs [hunk 1]
```diff
@@ -98,9 +106,9 @@ struct PerConnection<RpcMiddleware, HttpMiddleware> {
 }
 
 /// Start RPC server listening on given address.
-pub async fn start_server<M>(
+pub async fn start_server<M: Send + Sync + 'static>(
```
Comment (gui1117) [inline substrate/client/rpc-servers/src/lib.rs:109]: put it in the where clause or remove the where clause below.

Comment (nickvikeras) [reply substrate/client/rpc-servers/src/lib.rs]: Good catch. I had manually cherry-picked this code from our fork of the old substrate repo. Fixed it.

File: substrate/client/rpc-servers/src/lib.rs [hunk 2]
```diff
@@ -52,6 +52,14 @@ const MEGABYTE: u32 = 1024 * 1024;
 /// Type alias for the JSON-RPC server.
 pub type Server = jsonrpsee::server::ServerHandle;
 
+/// Type to encapsulate the server handle and listening address.
+pub struct ServerAndListenAddress {
```
Comment (niklasad1) [inline substrate/client/rpc-servers/src/lib.rs:56]: can you rename this new struct `Server` and remove the type-alias `Server` above?

File: substrate/client/service/src/builder.rs [hunk 3]
```diff
@@ -507,8 +508,21 @@ where
...
+	let listen_addrs = match listen_addr {
+		Some(socket_addr) => {
+			let mut multiaddr: Multiaddr = socket_addr.ip().into();
+			multiaddr.push(Protocol::Tcp(socket_addr.port()));
+			vec![multiaddr]
+		},
+		None => vec![],
+	};
+
+	let rpc_handlers = RpcHandlers {
```
Comment (niklasad1) [inline substrate/client/service/src/builder.rs:522]: It doesn't makes sense to add listen_addresses to `RpcHandlers` this just an in-memory way to perform rpc calls, please remove it RpcHandlers.

Why did you add it (maybe I'm missing something)?

File: substrate/client/rpc-servers/src/lib.rs [hunk 4]
```diff
@@ -52,6 +52,14 @@ const MEGABYTE: u32 = 1024 * 1024;
...
+	/// Handle to the rpc server
+	pub handle: Server,
```
Comment (niklasad1) [inline substrate/client/rpc-servers/src/lib.rs:58]: I would rather add getters for handle and listen_addr than to make this public to avoid breaking changes in the future....

File: substrate/client/service/src/lib.rs [hunk 5]
```diff
@@ -372,7 +382,7 @@ pub fn start_rpc_servers<R>(
 	config: &Configuration,
 	gen_rpc_module: R,
 	rpc_id_provider: Option<Box<dyn RpcSubscriptionIdProvider>>,
-) -> Result<Box<dyn std::any::Any + Send + Sync>, error::Error>
+) -> Result<(Box<dyn std::any::Any + Send + Sync>, Option<SocketAddr>), error::Error>
```
Comment (niklasad1) [inline substrate/client/service/src/lib.rs:385]: I'm fine with this.

But the return type is bit complicated and I don't remember why it's `Box<dyn Any>` but would be much nicer with a wrapper type that has an API to get the listen_addr instead returning a tuple.... because the Server type already has this information.

File: substrate/client/service/src/builder.rs (hunk 3 above)
Comment (nickvikeras) [reply substrate/client/service/src/builder.rs:522]: >Why did you add it (maybe I'm missing something)?

Just because the RpcHandlers type is what is getting returned to the caller. Would a new return type that can provide both the ```Server``` and the ```RpcHandlers``` be preferred?

Comment (niklasad1) [reply substrate/client/service/src/builder.rs:522]: Aight, I see one may want to know the listen addrs on the rpc endpoint after calling `spawn_tasks`.

It's probably fine to keep the `RpcHandlers` or rename to the RpcService or something, but it needs to documented that is represent a running RPC server as well with the legacy `in-memory rpc calls`.

We may want to remove latter at some point, it was for WASM stuff before smoldot was a thing...

File: substrate/client/rpc-servers/src/lib.rs (hunk 4 above)
Comment (nickvikeras) [reply substrate/client/rpc-servers/src/lib.rs]: Done.

File: substrate/client/service/src/builder.rs (hunk 3 above)
Comment (nickvikeras) [reply substrate/client/service/src/builder.rs:535]: I added some notes with a link to this comment thread.

File: substrate/client/rpc-servers/src/lib.rs (hunk 2 above)
Comment (nickvikeras) [reply substrate/client/rpc-servers/src/lib.rs]: Done.

File: substrate/client/service/src/lib.rs (hunk 5 above)
Comment (nickvikeras) [reply substrate/client/service/src/lib.rs]: I just used the Server type here, and moved the Drop wrapper up to the point where it was actually needed. Let me know if that suffices.

File: substrate/client/rpc-servers/src/lib.rs [hunk 6]
```diff
@@ -49,8 +50,31 @@ pub use middleware::{Metrics, MiddlewareLayer, RpcMetrics};
...
+}
+
+impl Server {
+	/// Creates a new Server.
+	pub fn new(handle: ServerHandle, listen_addr: Option<SocketAddr>) -> Server {
+		Server { handle, listen_addr }
+	}
+
+	/// Returns the `jsonrpsee::server::ServerHandle` for this Server. Can be used to perform
+	/// in-memory RPC queries.
```
Comment (niklasad1) [inline substrate/client/rpc-servers/src/lib.rs:68]:
```suggestion
```

File: substrate/client/rpc-servers/src/lib.rs [hunk 7]
```diff
@@ -49,8 +50,31 @@ pub use middleware::{Metrics, MiddlewareLayer, RpcMetrics};
...
+	listen_addr: Option<SocketAddr>,
+}
+
+impl Server {
+	/// Creates a new Server.
+	pub fn new(handle: ServerHandle, listen_addr: Option<SocketAddr>) -> Server {
+		Server { handle, listen_addr }
+	}
+
+	/// Returns the `jsonrpsee::server::ServerHandle` for this Server. Can be used to perform
```
Comment (niklasad1) [inline substrate/client/rpc-servers/src/lib.rs:67]:
```suggestion
	/// Returns the `jsonrpsee::server::ServerHandle` for this Server. Can be used to stop the server.
```

**Review by niklasad1 — APPROVED** [review]
nice work, looks good to me

File: substrate/client/service/src/lib.rs [hunk 8]
```diff
@@ -96,9 +98,17 @@ pub use task_manager::{SpawnTaskHandle, Task, TaskManager, TaskRegistry, DEFAULT
...
 #[derive(Clone)]
-pub struct RpcHandlers(Arc<RpcModule<()>>);
+pub struct RpcHandlers {
+	// This is legacy and may be removed at some point, it was for WASM stuff before smoldot was a
+	// thing. https://github.com/paritytech/polkadot-sdk/pull/5038#discussion_r1694971805
+	rpc_module: Arc<RpcModule<()>>,
+
+	// This can be used to introspect the port the RPC server is listening on. SDK consumers are
+	// depending on this and it should be supported even if in-memory query support is removed.
+	listen_addresses: Box<Vec<Multiaddr>>,
```
Comment (lexnv) [inline substrate/client/service/src/lib.rs:110]: dq: do we need the box here?

Comment (niklasad1) [reply substrate/client/service/src/lib.rs:110]: got catch, probably not :)

**Review by niklasad1 — APPROVED** [review]

**Comment by nickvikeras** [comment]
> hey @nickvikeras
> 
> This PR needs a prdoc description, [instructions how write and generate such here](https://github.com/paritytech/polkadot-sdk/blob/master/docs/contributor/prdoc.md) the label should be node dev. Can you provide such a thing?
> 
> I can help with that if you don't get it working.

I added it in the latest commit. Let me know if it looks good.

File: prdoc/pr_5038.prdoc [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+# See doc at https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/prdoc/schema_user.json
+
+title: Plumb RPC listener up to caller
+
+doc:
+  - audience: Node Dev
+    description:
+      This PR allows the RPC server's socket address to be returned when initializing the server.
+      This allows the library consumer to easily programmatically determine which port the RPC server is listening on.
+      My use case for this is automated testing. I'd like to be able to simply specify that the server bind to port '0'
```
Comment (niklasad1) [inline prdoc/pr_5038.prdoc:11]:
```suggestion
```

File: prdoc/pr_5038.prdoc [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+      and then test against whatever port the OS assigns dynamically. I will have many RPC servers running in parallel
```
Comment (niklasad1) [inline prdoc/pr_5038.prdoc:12]:
```suggestion
```

File: prdoc/pr_5038.prdoc [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+      across many tests within a single process, and I don't want to have to deal with port conflicts.
```
Comment (niklasad1) [inline prdoc/pr_5038.prdoc:13]:
```suggestion
```

File: prdoc/pr_5038.prdoc [hunk 12]
```diff
@@ -0,0 +1,18 @@
+# Schema: Polkadot SDK PRDoc Schema (prdoc) v1.0.0
+# See doc at https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/prdoc/schema_user.json
+
+title: Plumb RPC listener up to caller
+
+doc:
```
Comment (niklasad1) [inline prdoc/pr_5038.prdoc:6]: cool, I think the two first sentences are enough to describe this PR

**Review by niklasad1 — APPROVED** [review]

**Comment by nickvikeras** [comment]
I just pushed a merge commit with resolve a conflict with main.

**Review by niklasad1 — APPROVED** [review]

**Review by niklasad1 — APPROVED** [review]

**Comment by nickvikeras** [comment]
Gentle ping.

**Review by niklasad1 — APPROVED** [review]

**Comment by niklasad1** [comment]
Hey @nickvikeras 

Sorry, I merged #4792 and causes some merge conflicts that you have fix.
Let me know if you need some help with that.

**Review by niklasad1 — APPROVED** [review]

File: substrate/client/service/src/builder.rs [hunk 13]
```diff
@@ -372,6 +373,19 @@ pub struct SpawnTasksParams<'a, TBl: BlockT, TCl, TExPool, TRpc, Backend> {
 	pub telemetry: Option<&'a mut Telemetry>,
 }
 
+// Wrapper for HTTP and WS servers that makes sure they are properly shut down.
+mod waiting {
```
Comment (michalkucharczyk) [inline substrate/client/service/src/builder.rs:377]: From a quick glance, I don't see it is used anywhere? Is it acutally needed?

File: substrate/client/service/src/lib.rs [hunk 14]
```diff
@@ -451,7 +452,7 @@ where
 	match tokio::task::block_in_place(|| {
 		tokio_handle.block_on(sc_rpc_server::start_server(server_config))
 	}) {
-		Ok(server) => Ok(Box::new(waiting::Server(Some(server)))),
+		Ok(server) => Ok(server),
```
Comment (niklasad1) [inline substrate/client/service/src/lib.rs:455]: use waiting::Server(server) here
```suggestion
		Ok(server) => Ok(waiting::Server(server)),
```

**Review by michalkucharczyk — APPROVED** [review]

File: substrate/client/service/src/builder.rs (hunk 13 above)
Comment (niklasad1) [reply substrate/client/service/src/builder.rs:377]: good catch, added a suggestion for it

**Review by niklasad1 — CHANGES_REQUESTED** [review]
Hey again,

Just to fix the waiting stuff then it should be good to go, sorry missed that

**Review by niklasad1 — APPROVED** [review]

**Comment by nickvikeras** [comment]
Thank you!!! 🎉
